airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py

```diff
@@ -0,0 +1,78 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+from typing import Any, Dict, List, Optional, Union
+
+from airbyte_cdk.models import AirbyteRecordMessage
+from genson import SchemaBuilder
+from genson.schema.strategies.object import Object
+
+
+class NoRequiredObj(Object):
+    """
+    This class has Object behaviour, but it does not generate "required[]" fields
+    every time it parses object. So we dont add unnecessary extra field.
+    """
+
+    def to_schema(self):
+        schema = super(Object, self).to_schema()
+        schema["type"] = "object"
+        if self._properties:
+            schema["properties"] = self._properties_to_schema(self._properties)
+        if self._pattern_properties:
+            schema["patternProperties"] = self._properties_to_schema(self._pattern_properties)
+        return schema
+
+
+class NoRequiredSchemaBuilder(SchemaBuilder):
+    EXTRA_STRATEGIES = (NoRequiredObj,)
+
+# This type is inferred from the genson lib, but there is no alias provided for it - creating it here for type safety
+InferredSchema = Dict[str, Union[str, Any, List, List[Dict[str, Union[Any, List]]]]]
+
+
+class SchemaInferrer:
+    """
+    This class is used to infer a JSON schema which fits all the records passed into it
+    throughout its lifecycle via the accumulate method.
+
+    Instances of this class are stateful, meaning they build their inferred schemas
+    from every record passed into the accumulate method.
+
+    """
+
+    builders: Dict[str, SchemaBuilder]
+
+    def __init__(self):
+        self.builders = {}
```

Comment:
you can simplify the `if/else` block of `accumulate` by using `defaultdict` e.g: 

```
def __init__():
  self.builders = defaultdict(NoRequiredSchemaBuilder)

...

def accumulate(self, record): 
  self.builders[record.stream].add_object(record.data)

```

Reply:
Neat, didn't know about defaultdict